Client/Frontend/Sync/BraveCore/History/HistoryFetchers.swift

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveRewards
+import CoreData
+import OrderedCollections
+
+// MARK: - HistoryV2FetchResultsDelegate
+
+protocol HistoryV2FetchResultsDelegate: AnyObject {
+    
+    func controllerWillChangeContent(_ controller: HistoryV2FetchResultsController)
+    
+    func controllerDidChangeContent(_ controller: HistoryV2FetchResultsController)
+    
+    func controller(_ controller: HistoryV2FetchResultsController, didChange anObject: Any,
+                    at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?)
+    
+    func controller(_ controller: HistoryV2FetchResultsController, didChange sectionInfo: NSFetchedResultsSectionInfo,
+                    atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType)
+    
+    func controllerDidReloadContents(_ controller: HistoryV2FetchResultsController)
+}
+
+// MARK: - HistoryV2FetchResultsController
+
+protocol HistoryV2FetchResultsController {
+    
+    var delegate: HistoryV2FetchResultsDelegate? { get set }
+    
+    var fetchedObjects: [Historyv2]? { get }
+    
+    var fetchedObjectsCount: Int { get }
+    
+    var sectionCount: Int { get }
+    
+    func performFetch(_ completion: @escaping () -> Void)
+    
+    func object(at indexPath: IndexPath) -> Historyv2?
+    
+    func objectCount(for section: Int) -> Int
+    
+    func titleHeader(for section: Int) -> String
+
+}
+
+// MARK: - Historyv2Fetcher
+
+class Historyv2Fetcher: NSObject, HistoryV2FetchResultsController {
+    
+    // MARK: Lifecycle
+    
+    init(historyAPI: BraveHistoryAPI) {
+        self.historyAPI = historyAPI
+        super.init()
+        
+        self.historyServiceListener = historyAPI.add(HistoryServiceStateObserver { [weak self] _ in
+            guard let self = self else { return }
+            
+            DispatchQueue.main.async {
+                self.delegate?.controllerDidReloadContents(self)
+            }
+        })
+    }
+    
+    // MARK: Internal
+    
+    weak var delegate: HistoryV2FetchResultsDelegate?
+    
+    var fetchedObjects: [Historyv2]? {
+        historyList
+    }
+    
+    var fetchedObjectsCount: Int {
+        historyList.count
+    }
+    
+    var sectionCount: Int {
+        return sectionDetails.elements.filter { $0.value > 0 }.count
+    }
+    
+    func performFetch(_ completion: @escaping () -> Void) {
+        clearHistoryData()
+        
+        historyAPI?.search(withQuery: "", maxCount: 0, completion: { [weak self] historyNodeList in
```

Comment:
`maxCount = 0` means no there's no fetch limit? Is this performant enough?

Reply:
Great catch, this has be "200" like we are doing before.

Fix at https://github.com/brave/brave-ios/pull/3703/commits/859375beafc1a0b1778a58dd79cb21c16bacd6cf